compiler/compiler/src/run_with_ledger.rs

```
@@ -0,0 +1,299 @@
+    }
+
+    let mut case_outcomes = Vec::new();
+
+    for case in cases {
+        assert!(
+            ledger.vm().contains_program(&ProgramID::from_str(&case.program_name).unwrap()),
+            "Program {} should exist.",
+            case.program_name
+        );
+
+        let private_key = case
+            .private_key
+            .as_ref()
+            .map(|key| PrivateKey::from_str(key).expect("Failed to parse private key."))
+            .unwrap_or(genesis_private_key);
+
+        let mut execution = None;
+        let mut verified = false;
+        let mut status = Status::None;
+
+        // Halts are handled by panics, so we need to catch them.
+        // I'm not thrilled about this usage of `AssertUnwindSafe`, but it seems to be
+        // used frequently in SnarkVM anyway.
+        let execute_output = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+            ledger.vm().execute(
+                &private_key,
+                (&case.program_name, &case.function),
+                case.input.iter(),
+                None,
+                0,
+                None,
+                &mut rng,
+            )
+        }));
+
+        if let Err(payload) = execute_output {
+            let s1 = payload.downcast_ref::<&str>().map(|s| s.to_string());
+            let s2 = payload.downcast_ref::<String>().cloned();
+            let s = s1.or(s2).unwrap_or_else(|| "Unknown panic payload".to_string());
+
+            case_outcomes.push(CaseOutcome {
+                status: Status::Halted(s),
+                verified: false,
+                errors: buf.extract_errs(),
+                warnings: buf.extract_warnings(),
+                execution: "".to_string(),
+            });
+            continue;
+        }
+
+        let result = execute_output
+            .unwrap()
+            .and_then(|transaction| {
+                verified = ledger.vm().check_transaction(&transaction, None, &mut rng).is_ok();
```

Would we want to change `verified` to a `String` so that we can bubble up the error message?